src/hotspot/cpu/ppc/stubGenerator_ppc.cpp

```diff
@@ -3544,6 +3544,461 @@ class StubGenerator: public StubCodeGenerator {
+    __ load_const_optimized(const_ptr, (address)&bitposLUT_val, tmp_reg);
+    __ lxv(bitposLUT, 0, const_ptr);
+    if (PowerArchitecturePPC64 >= 10) {
+        __ load_const_optimized(const_ptr, (address)&p10_pack_permute_val, tmp_reg);
+    } else {
+        __ load_const_optimized(const_ptr, (address)&pack_rshift_val, tmp_reg);
+        __ lxv(pack_rshift->to_vsr(), 0, const_ptr);
+        __ load_const_optimized(const_ptr, (address)&pack_lshift_val, tmp_reg);
+        __ lxv(pack_lshift->to_vsr(), 0, const_ptr);
+        __ load_const_optimized(const_ptr, (address)&pack_permute_val, tmp_reg);
+    }
+    __ lxv(pack_permute, 0, const_ptr);
+
+    // Splat the constants that can use xxspltib
+    __ xxspltib(vec_0s->to_vsr(), 0);
+    __ xxspltib(vec_4s->to_vsr(), 4);
+    __ xxspltib(vec_8s->to_vsr(), 8);
+    __ xxspltib(vec_0xfs, 0xf);
+    if (PowerArchitecturePPC64 >= 10) {
+        __ xxspltib(vec_0x3fs->to_vsr(), 0x3f);
+    }
+
+    // The rest of the constants use different values depending on the
+    // setting of isURL
+    __ cmpwi(CCR0, isURL, 0);
+    __ beq(CCR0, not_URL);
+
+    // isURL != 0 (true)
+    __ load_const_optimized(const_ptr, (address)&offsetLUT_URL_val, tmp_reg);
+    __ lxv(offsetLUT, 0, const_ptr);
+    __ load_const_optimized(const_ptr, (address)&maskLUT_URL_val, tmp_reg);
+    __ lxv(maskLUT, 0, const_ptr);
+    __ xxspltib(vec_special_case_char->to_vsr(), '_');
+    __ xxspltib(vec_special_case_offset, (unsigned char)US);
+    __ b(calculate_size);
+
+    // isURL = 0 (false)
+    __ bind(not_URL);
+    __ load_const_optimized(const_ptr, (address)&offsetLUT_val, tmp_reg);
+    __ lxv(offsetLUT, 0, const_ptr);
+    __ load_const_optimized(const_ptr, (address)&maskLUT_val, tmp_reg);
+    __ lxv(maskLUT, 0, const_ptr);
+    __ xxspltib(vec_special_case_char->to_vsr(), '/');
+    __ xxspltib(vec_special_case_offset, (unsigned char)SLS);
+
+    __ bind(calculate_size);
+
+    // Don't handle the last 4 characters of the source, because this
+    // VSX-based algorithm doesn't handle padding characters.  Also the
+    // vector code will always write 16 bytes of decoded data on each pass,
+    // but only the first 12 of those 16 bytes are valid data (16 base64
+    // characters become 12 bytes of binary data), so for this reason we
+    // need to subtract an additional 8 bytes from the source length, in
+    // order not to write past the end of the destination buffer.  The
+    // result of this subtraction implies that the non-instrinsic routine
+    // will be used to process the last 12 characters.
+    __ sub(sl, sl, sp);
+    __ subi(sl, sl, 12);
+
+    // Round sl down to the nearest multiple of block_size
+    __ clrrdi(sl, sl, block_size_clear);
+
+    // out starts at d + dp
+    __ add(out, d, dp);
+
+    // in starts at s + sp
+    __ add(in, s, sp);
+
+    // Address of the last byte of the source is (in + sl - 1)
+    __ add(end, in, sl);
+    __ subi(end, end, 1);
```

Comment:
Looks a bit complicated, but ok. I'd have loaded the number of iterations into ctr and used bdnz instruction for the loop.

Reply:
I agree that a count would be preferable, because it's easier to read.  Will fix.